cpp/src/codegen/arrow_compute/ext/sort_kernel.cc

```diff
@@ -2259,27 +2280,41 @@ extern "C" void MakeCodeGen(arrow::compute::FunctionContext* ctx,
     std::stringstream ss;
     for (int i = 0; i < key_index_list_.size(); i++) {
       ss << "cached_" << i << "_.push_back(std::make_shared<ArrayType_" << i 
-         << ">(in[" << i << "]));" << std::endl;
+         << ">(in[" << i << "]));\n"
+         // update has_null_
+         << "if (!has_null_ && cached_" << i << "_[cached_" << i 
+         << "_.size() - 1]->null_count() > 0) {"
+         << "has_null_ = true;}" << std::endl;
     }
     return ss.str();
   }
   std::string GetSortFunction() {
-    return "gfx::timsort(indices_begin, indices_begin + "
-           "items_total_, "
-           "comp);";
+    std::stringstream ss;
+    ss << "if (has_null_) {\n"
+       << "gfx::timsort(indices_begin, indices_begin + items_total_, comp);} else {\n" 
+       << "gfx::timsort(indices_begin, indices_begin + items_total_, comp_without_null);}"
+       << std::endl;
+    return ss.str();
+
   }
-  std::string GetCompFunction() {
+  std::string GetCompFunction(bool has_null) {
     std::stringstream ss;
     bool projected;
     if (key_projector_) {
       projected = true;
     } else {
       projected = false;
     }
-    ss << "auto comp = [this](ArrayItemIndexS x, ArrayItemIndexS y) {"
-       << GetCompFunction_(0, projected, key_field_list_,
-                           projected_types_, sort_directions_, nulls_order_)
-       << "};";
+    if (has_null) {
+      ss << "auto comp = [this](ArrayItemIndexS x, ArrayItemIndexS y) {"
+         << GetCompFunction_(0, projected, key_field_list_,
+                             projected_types_, sort_directions_, nulls_order_);
+    } else {
+      ss << "auto comp_without_null = [this](ArrayItemIndexS x, ArrayItemIndexS y) {"
```

Comment:
should pass by reference here
`[this](const ArrayItemIndexS& x, const ArrayItemIndexS& y)`

Reply:
As below error occurred for timsort, so only replaced this in std sort.
`timsort.hpp:93:16: error: binding reference of type ‘sparkcolumnarplugin::codegen::arrowcompute::extra::ArrayItemIndexS&’ to ‘const sparkcolumnarplugin::codegen::arrowcompute::extra::ArrayItemIndexS’ discards qualifiers
         return static_cast<bool>(compare(
                ^~~~~~~~~~~~~~~~~~~~~~~~~~
             projection(std::forward<T>(lhs)),
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
             projection(std::forward<U>(rhs))
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
         ));`